### PR TITLE
shuffle and randomly drop tags in lora training script

### PR DESCRIPTION
**Pull request by noskill**
shuffle and drop tags to decrease overfitting on small datasets

## Who can review?

@sayakpaul

### CODE REVIEW

**Comment by bghira** [comment]
* please try to run `make style` to fix formatting errors
* please switch the `-` in the parameter name to `_` to match convention, eg. `--pretrained_model_name_or_path`
* you can use `choices=linspace(0, 1)` and `metavar="[0-1]"` to state inside `add_argument` that it has a float range instead of a check after parsing, which puts proper `--help` output as well
* introducing this logic makes it impossible to also meaningfully support cached text embeds
* maybe it should be added to the other training examples, too

**Comment by noskill** [comment]
@bghira do you mean numpy.linspace?

File: examples/text_to_image/train_text_to_image_lora_sdxl.py [hunk 1]
```diff
@@ -431,6 +431,16 @@ def parse_args(input_args=None):
         help="debug loss for each image, if filenames are awailable in the dataset",
     )
 
+    parser.add_argument(
+        "--shuffle_tags",
+        action="store_true",
+        help="shuffle tags randomly")
```
Comment (bghira) [inline examples/text_to_image/train_text_to_image_lora_sdxl.py:437]: it seems like 'make style' didn't work, as the ) is on the same line as help

can you rephrase the help statements to have Capitalised initial letter, and a bit more explanation about when shuffle tags might be used?

**Comment by noskill** [comment]
with linspace it too easy to get invalid choice error with thousands of numbers in output

**Comment by bghira** [comment]
there should really just be 100 choices though, as it's a percentage

**Comment by noskill** [comment]
Do you mean p_drop_tags should be accept integer instead of floating point value? 
With choices=torch.linspace(0, 100, 100, dtype=torch.int8) i can pass 21, but with torch.linspace(0, 1, 100)  i can't pass 0.21

**Comment by sayakpaul** [comment]
Thanks for your contributions. Could you demonstrate this with some results? Will be very helpful for us to prioritize then. Otherwise, I think it might be best to just include a note about this technique in the README.

**Comment by noskill** [comment]
I have proof that changing tag order changes generated image:

```
from diffusers import AutoPipelineForText2Image
from diffusers.schedulers import DPMSolverMultistepScheduler

import torch

model_name = 'frankjoshua/juggernautXL_v8Rundiffusion'
pipeline = AutoPipelineForText2Image.from_pretrained(model_name, torch_dtype=torch.float16,  custom_pipeline="lpw_stable_diffusion_xl").to("cuda")

pipeline.scheduler = DPMSolverMultistepScheduler.from_config(pipeline.scheduler.config)

pipeline.vae.enable_tiling()

prompt = "farmer is holding his sheep, farm, smart, best quality, camera low, camera close, best quality, amazing,ultra high res, masterpiece"
negative_prompt = "(deformed, distorted, disfigured:1.3), poorly drawn, bad anatomy, wrong anatomy, extra limb, missing limb, floating limbs, (mutated hands and fingers:1.4), disconnected limbs, mutation, mutated, ugly, disgusting, blurry, amputation"

generator = torch.Generator('cuda').manual_seed(4394893432342354523)
images = [pipeline(prompt, num_inference_steps=25, generator=generator, negative_prompt=negative_prompt, guidance_scale=5, cross_attention_kwargs={"scale": 0.99}).images[0] for _ in range(10)]

i = 0
for im in images:
    im.save(f'farms/farm{i}.jpeg')
    i += 1
```

changing  places of "farm" and "smart":
farm9
farm91


Though i don't know yet how much effect it has on training.

**Comment by sayakpaul** [comment]
Thanks for showing the example. I think it might be just better to add a section about this technique along with your results.

File: examples/text_to_image/README_sdxl.md [hunk 2]
```diff
@@ -182,7 +182,8 @@ The above command will also run inference as fine-tuning progresses and log the
 **Notes**:
 
 * SDXL's VAE is known to suffer from numerical instability issues. This is why we also expose a CLI argument namely `--pretrained_vae_model_name_or_path` that lets you specify the location of a better VAE (such as [this one](https://huggingface.co/madebyollin/sdxl-vae-fp16-fix)).
-
+* there are options that control data augmentation on text level `--shuffle_tags` and `--p_drop_tags`, it might be worth
+  turning them on a small dataset.
```
Comment (bghira) [inline examples/text_to_image/README_sdxl.md:186]: 'enabling on a small dataset'

File: examples/text_to_image/README_sdxl.md [hunk 3]
```diff
@@ -182,7 +182,8 @@ The above command will also run inference as fine-tuning progresses and log the
 **Notes**:
 
 * SDXL's VAE is known to suffer from numerical instability issues. This is why we also expose a CLI argument namely `--pretrained_vae_model_name_or_path` that lets you specify the location of a better VAE (such as [this one](https://huggingface.co/madebyollin/sdxl-vae-fp16-fix)).
-
+* there are options that control data augmentation on text level `--shuffle_tags` and `--p_drop_tags`, it might be worth
```
Comment (bghira) [inline examples/text_to_image/README_sdxl.md:185]: "T"here (casing)

File: examples/text_to_image/train_text_to_image_lora.py [hunk 4]
```diff
@@ -188,6 +188,21 @@ def parse_args():
         default="sd-model-finetuned-lora",
         help="The output directory where the model predictions and checkpoints will be written.",
     )
+    parser.add_argument(
+        "--debug_loss",
+        action="store_true",
+        help="debug loss for each image, if filenames are awailable in the dataset",
+    )
```
Comment (bghira) [inline examples/text_to_image/train_text_to_image_lora.py:195]: not sure this is supposed to be here

**Comment by yiyixuxu** [comment]
cc @sayakpaul here

**Comment by sayakpaul** [comment]
Again, thanks for working on this. But I think we agreed on keeping it this way?
https://github.com/huggingface/diffusers/pull/7467#issuecomment-2022533004

By that, I mean we include sections in our REAMDE and show a small code snippet showing how to implement it. Would that work for you?

**Comment by noskill** [comment]
@sayakpaul I thought what i added to readme is enough, do you mean i need to add a code snippet showing how shuffling tags is implemented?
